modules/tensor_mechanics/src/materials/ComputeCappedWeakPlaneStress.C

```diff
@@ -0,0 +1,890 @@
+  _elastic_strain[_qp] = _elastic_strain_old[_qp] + _strain_increment[_qp] - (_plastic_strain[_qp] - _plastic_strain_old[_qp]);
+
+  //Rotate the tensors to the current configuration
+  if (_perform_finite_strain_rotations)
+  {
+    _stress[_qp] = _rotation_increment[_qp]*_stress[_qp]*_rotation_increment[_qp].transpose();
+    _elastic_strain[_qp] = _rotation_increment[_qp] * _elastic_strain[_qp] * _rotation_increment[_qp].transpose();
+    _plastic_strain[_qp] = _rotation_increment[_qp] * _plastic_strain[_qp] * _rotation_increment[_qp].transpose();
+  }
+}
+
+void
+ComputeCappedWeakPlaneStress::returnMap()
+{
+  // initially assume an elastic deformation
+  _stress[_qp] = _stress_old[_qp] + _elasticity_tensor[_qp] * _strain_increment[_qp];
+  _intnl[_qp][0] = _intnl_old[_qp][0];
+  _intnl[_qp][1] = _intnl_old[_qp][1];
+  _iter[_qp] = 0.0;
+  _linesearch_needed[_qp] = 0.0;
+
+  _stress_return_type = nothing_special;
+
+  Real p_trial = _stress[_qp](2, 2);
+  Real q_trial = std::sqrt(Utility::pow<2>(_stress[_qp](2, 0)) + Utility::pow<2>(_stress[_qp](2, 1)));
+  unsmoothedYieldFunctions(p_trial, q_trial, _intnl[_qp], _yf[_qp]);
+
+  /* Need to consider smoothing_tol, not just yf<=0, because
+   * some yield functions might mix.  However, if some yield
+   * functions are -smoothing_tol <= yf < 0, then the smoothed
+   * yield function must be computed and checked vs f_tol
+   */
+  if ((_yf[_qp][0] < -_smoothing_tol &&
+      _yf[_qp][1] < -_smoothing_tol &&
+      _yf[_qp][2] < -_smoothing_tol) ||
+      yieldF(p_trial, q_trial, _intnl[_qp]) <= _f_tol)
+  {
+    // elastic case
+    _plastic_strain[_qp] = _plastic_strain_old[_qp];
+    if (_fe_problem.currentlyComputingJacobian())
+      _Jacobian_mult[_qp] = _elasticity_tensor[_qp];
+    return;
+  }
+
+
+  /* The trial stress (_stress[_qp]) must be inadmissible
+   * so we need to return to the yield surface.  The following
+   * equations must be satisfied.
+   * 0 = f(p, q, intnl)                    = rhs[0]
+   * 0 = p - p^trial + Ezzzz * ga * dg/dp  = rhs[1]
+   * 0 = q - q^trial + Ezxzx * ga * dg/dq  = rhs[2]
+   * 0 = intnl[0] - intnl_old[0] - (q^trial - q)/Ezxzx
+   * 0 = intnl[1] - intnl_old[1] - (p^trial - p)/Ezzzz + (q^trial - q) * tanpsi / Ezxzx
+   * With unknowns p, q, ga, intnl[0] and intnl[1]
+   * I find it convenient to solve the first three equations for p, q and ga*Ezzzz,
+   * while substituting the last two equations into these during the solve process
+   * (tanpsi depends only on intnl[0], and hence on q)
+   */
+
+  const Real in_trial20 = _stress[_qp](2, 0);
+  const Real in_trial21 = _stress[_qp](2, 1);
+  const Real in_q_trial = q_trial;
+
+  // values of p, q  and intnl we know to be admissible
+  Real p_ok = _stress_old[_qp](2, 2);
+  Real q_ok = std::sqrt(Utility::pow<2>(_stress[_qp](2, 0)) + Utility::pow<2>(_stress[_qp](2, 1)));
+  // note that _intnl[_qp] is the "running" intnl variable that gets updated every NR iteration
+  std::vector<Real> intnl_ok(2);
```

Comment:
std::array with initialization list here: Should be much faster (no allocation/deallocation) and looks nice:
```
std::array<Real, 2> initnl_ok = { _intnl_old[_qp][0], _intnl_old[_qp][1] };
```